### PR TITLE
[ie/facebook] add handling of facebook.com/events

### LINKED ISSUE #5355: facebook extractor - added handling of facebook.com/event/id
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.10.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Italy

### Provide a description that is worded well enough to be understood

Facebook video pages of the form
https://www.facebook.com/events/
do not get downloaded.

I made a fix for it which is in the attachment
[yt-dlp-20221025.1.diff.txt](https://github.com/yt-dlp/yt-dlp/files/9861917/yt-dlp-20221025.1.diff.txt)

which is:
```
--- /home/mar/PycharmProjects/yt-dlp/yt-dlp/venv/lib/python3.9/site-packages/yt_dlp/extractor/facebook.py
+++ /home/mar/PycharmProjects/yt-dlp/#backup/yt-dlp/venv/lib/python3.9/site-packages/yt_dlp/extractor/facebook.py
@@ -50,6 +50,7 @@
                             )\?(?:.*?)(?:v|video_id|story_fbid)=|
                             [^/]+/videos/(?:[^/]+/)?|
                             [^/]+/posts/|
+                            events/(?:[^/]+/)?|
                             groups/[^/]+/permalink/|
                             watchparty/
                         )|
@@ -562,6 +563,15 @@
                     if not entries:
                         parse_graphql_video(video)

+                if not entries:
+                    event = data.get('event') or {}
+                    if event:
+                        cover_media_renderer = event.get('cover_media_renderer') or {}
+                        if cover_media_renderer:
+                            cover_video = cover_media_renderer.get('cover_video') or {}
+                            if cover_video:
+                                parse_graphql_video(cover_video)
+
                 if len(entries) > 1:
                     return self.playlist_result(entries, video_id)
```

I hope it helps. I prefer not to commit the patch myself.

I'd have one question.
- the path ['online_cta_renderer']['event']['live_virtual_event_info']['associated_live_videos']['edges']
  gives me a 'node' from 'canonical_url_with_fallback' which gives me an url that if I download manually with yt-dlp would give 
  me a better thumbnail. This is not important, but generally speaking, is there a way to do the same automatically from the function
  `def _extract_from_url(self, url, video_id)` ?

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
/home/mar/PycharmProjects/yt-dlp/yt-dlp/venv/bin/python /home/mar/PycharmProjects/yt-dlp/yt-dlp/DownloadVideo.py
No module named 'selenium'
[debug] print_options: True
[debug] writeurlshortcut: True
[debug] zipsubtitles: True
[debug] writesubtitles: True
[debug] writeautomaticsub: True
[debug] allsubtitlesnormal: True
[debug] skip_download: False
[debug] subtitlesformat: srt/best
[debug] subtitleslangs: ['en', 'it', 'es', 'fr', 'pt', 'de']    ***
[debug] subtitles_converter: <function subtitles_converter at 0x7f8d6cdf3e50>
[debug] subtitles_converted_overwrite_destination: True
[debug] subtitles_converted_keep_source: True
[debug] writethumbnail: True
[debug] writelink: True    ***
[debug] writeurllink: True    ***
[debug] nooverwrites: True
[debug] writeannotations: True
[debug] postprocessor_pages: <function postprocessor_pages at 0x7f8d6519e550>
[debug] continue_dl: True
[debug] no_cache_dir: True
[debug] rm_cache_dir: True
[debug] format: best[height=460]/best[height>360]/best    ***
[debug] basePath: ~/work/files/
[debug] outtmpl: {'default': '~/work/files/videos/%(title)s/%(title)s - %(uploader)s-%(channel)s - %(playlist)s-%(playlist_index)s-%(id)s - %(extractor)s - %(format_id)s-%(height)sp-%(fps)sfps-%(asr)s.%(ext)s', 'chapter': '%(title)s - %(section_number)03d %(section_title)s [%(id)s].%(ext)s'}
[debug] outtmpl_video_files: ~/work/files/videos_pages/%(title)s/%(title)s - %(format_id)s-%(height)sp-%(fps)sfps-%(asr)s.%(ext)s
[debug] outtmpl_pages: ~/work/files/videos_pages/%(title)s/%(title)s.%(ext)s
[debug] max_dir_length: 250    ***
[debug] max_file_length: 232
[generic] 1509582499515440: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] 1509582499515440: Extracting information
ERROR: Unsupported URL: https://m.facebook.com/events/1509582499515440
Traceback (most recent call last):
  File "/home/mar/PycharmProjects/yt-dlp/yt-dlp/venv/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 1478, in wrapper
    return func(self, *args, **kwargs)
  File "/home/mar/PycharmProjects/yt-dlp/yt-dlp/venv/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 1554, in __extract_info
    ie_result = ie.extract(url)
  File "/home/mar/PycharmProjects/yt-dlp/yt-dlp/venv/lib/python3.9/site-packages/yt_dlp/extractor/common.py", line 674, in extract
    ie_result = self._real_extract(url)
  File "/home/mar/PycharmProjects/yt-dlp/yt-dlp/venv/lib/python3.9/site-packages/yt_dlp/extractor/generic.py", line 2756, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://m.facebook.com/events/1509582499515440

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mar/PycharmProjects/yt-dlp/yt-dlp/DownloadVideo.py", line 706, in <module>
    main()
  File "/home/mar/PycharmProjects/yt-dlp/yt-dlp/DownloadVideo.py", line 613, in main
    download_video(ydl_opts, urls)
  File "/home/mar/PycharmProjects/yt-dlp/yt-dlp/DownloadVideo.py", line 697, in download_video
    ydl.download(urls)           # vids
  File "/home/mar/PycharmProjects/yt-dlp/yt-dlp/venv/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 3316, in download
    self.__download_wrapper(self.extract_info)(
  File "/home/mar/PycharmProjects/yt-dlp/yt-dlp/venv/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 3291, in wrapper
    res = func(*args, **kwargs)
  File "/home/mar/PycharmProjects/yt-dlp/yt-dlp/venv/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 1467, in extract_info
    return self.__extract_info(url, self.get_info_extractor(key), download, extra_info, process)
  File "/home/mar/PycharmProjects/yt-dlp/yt-dlp/venv/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 1496, in wrapper
    self.report_error(str(e), e.format_traceback())
  File "/home/mar/PycharmProjects/yt-dlp/yt-dlp/venv/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 997, in report_error
    self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
  File "/home/mar/PycharmProjects/yt-dlp/yt-dlp/venv/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 937, in trouble
    raise DownloadError(message, exc_info)
yt_dlp.utils.DownloadError: ERROR: Unsupported URL: https://m.facebook.com/events/1509582499515440

Process finished with exit code 1
```

### PR DESCRIPTION
**Pull request by kclauhk**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Add handling of facebook.com/events

Fixes #5355 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by bashonly** [comment]
Please don't combine several unrelated issues into 1 PR. Separate PRs are preferable. There are also changes that seem unrelated to the 4 issues, please explain why they are necessary or revert

**Comment by kclauhk** [comment]
> Please don't combine several unrelated issue fixes into 1 PR. Separate PRs are preferable. There are also changes that seem unrelated to the 4 issues, please explain why they are necessary or revert

Except the updates of some 'test' data, please kindly specify which changes seems unrelated to the 4 issues,

File: yt_dlp/extractor/facebook.py [hunk 1]
```diff
@@ -495,11 +498,15 @@ def extract_metadata(webpage):
             description = description or self._html_search_meta(
                 ['description', 'og:description', 'twitter:description'],
                 webpage, 'description', default=None)
+            uploader_data = (
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:501]: why was this moved

File: yt_dlp/extractor/facebook.py [hunk 2]
```diff
@@ -616,6 +625,32 @@ def parse_graphql_video(video):
                                 'url': playable_url,
                             })
                     extract_dash_manifest(video, formats)
+
+                    automatic_captions, subtitles = {}, {}
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:629]: why was this moved?

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/facebook.py (hunk 1 above)
Comment (kclauhk) [reply yt_dlp/extractor/facebook.py:501]: first, 'or get_first(post, ('event', 'event_creator', {dict})) or {})' added
second, move the 'uploader_data' block right before the 'uploader = ' seems more easier to read and maintain afterward

File: yt_dlp/extractor/facebook.py (hunk 2 above)
Comment (kclauhk) [reply yt_dlp/extractor/facebook.py:629]: to avoid complicated traverse_obj in
'''
460            subs_data = traverse_obj(post, (..., 'video', ..., 'attachments', ..., lambda k, v: (
461                k == 'media' and str(v['id']) == video_id and v['__typename'] == 'Video')))
'''
because events use different keys.
Moving subtitle extraction inside 'parse_graphql_video()' then we can remove these two lines

Comment (bashonly) [reply yt_dlp/extractor/facebook.py:607]: Okay that makes sense. But moving the whole code block made the diff much less obvious, and when there were 3 other unrelated changes in 1 PR I didn't even know what to look for. So thank you for separating the PRs

File: yt_dlp/extractor/facebook.py (hunk 1 above)
Comment (bashonly) [reply yt_dlp/extractor/facebook.py:483]: Makes sense. Same as my other comment though, moving the whole code block made the diff harder to parse

File: yt_dlp/extractor/facebook.py [hunk 3]
```diff
@@ -616,6 +603,32 @@ def parse_graphql_video(video):
...
+                    is_video_broadcast = get_first([video], 'is_video_broadcast', expected_type=bool)
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:608]: `get_first(x, y)` is just `traverse_obj(x, (..., y), get_all=False)`
```suggestion
                    is_video_broadcast = traverse_obj(video, ('is_video_broadcast', {bool}))
```

File: yt_dlp/extractor/facebook.py [hunk 4]
```diff
@@ -616,6 +603,32 @@ def parse_graphql_video(video):
...
+                    if url_or_none(captions):  # if video only had a 'captions_url'
+                        locale = self._html_search_meta(['og:locale', 'twitter:locale'], webpage, 'locale', default='en_US')
+                        auto_gen = (traverse_obj(video,
+                                    ('video_available_captions_locales', ..., 'localized_creation_method'), get_all=False)
+                                    or is_video_broadcast)
+                        (automatic_captions if auto_gen else subtitles)[locale] = [{'url': captions}]
+                    # or else video had 'video_available_captions_locales', a list of dicts
+                    for caption in traverse_obj(captions, (
+                        {lambda x: sorted(x, key=lambda c: c['locale'])}, lambda _, v: v['captions_url'])
+                    ):
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:621]: we don't have use the same var for the single `captions_url` and the `captions` list anymore
```suggestion
                    if captions_url := traverse_obj(video, ('captions_url', {url_or_none})):
                        locale = self._html_search_meta(
                            ['og:locale', 'twitter:locale'], webpage, 'locale', default='en_US')
                        auto_gen = is_video_broadcast or traverse_obj(video, (
                            'video_available_captions_locales', ..., 'localized_creation_method'))
                        (automatic_captions if auto_gen else subtitles)[locale] = [{'url': captions_url}]
                    for caption in traverse_obj(video, (
                        'video_available_captions_locales',
                        {lambda x: sorted(x, key=lambda c: c['locale'])},
                        lambda _, v: url_or_none(v['captions_url'])
                    )):
```

File: yt_dlp/extractor/facebook.py [hunk 5]
```diff
@@ -659,7 +674,8 @@ def parse_attachment(attachment, key='media'):
                 for edge in edges:
                     parse_attachment(edge, key='node')
 
-                video = data.get('video') or {}
+                video = (traverse_obj(data, ('event', 'cover_media_renderer', 'cover_video', {dict}))  # event cover video
+                         or data.get('video') or {})
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:678]: it's not good to do `traverse_obj(data, ...) or data.get(...)` when we can pass multiple paths to `traverse_obj` instead
```suggestion
                video = traverse_obj(data, (
                    'event', 'cover_media_renderer', 'cover_video'), 'video', expected_type=dict) or {}
```

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/facebook.py (hunk 4 above)
Comment (kclauhk) [reply yt_dlp/extractor/facebook.py:621]: For this subtitle part, the source dict feeded into `parse_graphql_video()` looks like this:

case 1:
only one subtitle (exception: FB Events give us incomplete value in `video_available_captions_locales` instead of `[]`)
```
{
    ...,
    'captions_url': 'https://...',
    'video_available_captions_locales': [],
    ...,
}
```
case 2: multiple subtitles:
```
{
    ...,
    'captions_url': 'https://...',
    'video_available_captions_locales': [
        {
            'localized_creation_method': None or 'Auto-generated',
            'captions_url': 'https://...',
            'locale': '?????',
            ...,
        }, {
            'localized_creation_method': None or 'Auto-generated',
            'captions_url': 'https://...',
            'locale': '?????',
        },
        ...,
    ],
    ...,
}
```
The value in `captions_url` (the outer one) is one of the url in `video_available_captions_locales` depended on locale (locale can be set via cookies). Therefore, ```captions = get_first(subs_data, 'video_available_captions_locales', 'captions_url')``` exists in the original codes in order to pick either `video_available_captions_locales` or `captions_url` (outer) with non-empty `video_available_captions_locales` in 1st priority (but FB Events make this line fail).

Regarding the suggestion, we will get this in case 2:
```
{
    'en_US': [
        {
            'url': 'https...'   # came from 'captions_url' (outer) when locale='en_US' ('en_US' from line 610)
        },
        {
            'url': 'https...',  # came from 'video_available_captions_locales' where 'locale'='en_US'
            'name': 'English'
        }
    ]
}
```
because it effectively picks both of the two.

Comment (bashonly) [reply yt_dlp/extractor/facebook.py:621]: then let's do this
```suggestion
                    for caption in traverse_obj(video, (
                        'video_available_captions_locales',
                        {lambda x: sorted(x, key=lambda c: c['locale'])},
                        lambda _, v: url_or_none(v['captions_url'])
                    )):
                        lang = caption.get('localized_language', '')
                        subs = {
                            'url': caption['captions_url'],
                            'name': format_field(caption, 'localized_country', f'{lang} (%s)', default=lang),
                        }
                        if caption.get('localized_creation_method') or is_video_broadcast:
                            automatic_captions.setdefault(caption['locale'], []).append(subs)
                        else:
                            subtitles.setdefault(caption['locale'], []).append(subs)
                    else:
                        if captions_url := traverse_obj(video, ('captions_url', {url_or_none})):
                            locale = self._html_search_meta(
                                ['og:locale', 'twitter:locale'], webpage, 'locale', default='en_US')
                            (automatic_captions if is_video_broadcast else subtitles)[locale] = [{'url': captions_url}]
```

File: yt_dlp/extractor/facebook.py [hunk 6]
```diff
@@ -616,6 +603,30 @@ def parse_graphql_video(video):
...
+                        if caption.get('localized_creation_method') or is_broadcast:
+                            automatic_captions.setdefault(caption['locale'], []).append(subs)
+                        else:
+                            subtitles.setdefault(caption['locale'], []).append(subs)
+                    else:
+                        if not automatic_captions and not subtitles:
+                            if captions_url := traverse_obj(video, ('captions_url', {url_or_none})):
+                                locale = self._html_search_meta(
+                                    ['og:locale', 'twitter:locale'], webpage, 'locale', default='en_US')
+                                (automatic_captions if is_broadcast else subtitles)[locale] = [{'url': captions_url}]
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:628]: oops, else was not what we wanted here, mb
```suggestion
                    captions_url = traverse_obj(video, ('captions_url', {url_or_none}))
                    if captions_url and not automatic_captions and not subtitles:
                            locale = self._html_search_meta(
                                ['og:locale', 'twitter:locale'], webpage, 'locale', default='en_US')
                            (automatic_captions if is_broadcast else subtitles)[locale] = [{'url': captions_url}]
```

**Review by bashonly — APPROVED** [review]

**Comment by frandroid** [comment]
Screenshot 2024-03-08 at 17 55 36

Now THAT'S obfuscation...  The blob should be easy to isolate though!

**Comment by kclauhk** [comment]
Video data is not extracted from Video tag therefore the blob here is irrelevent.
Please provide us the url of Fb event (https://www.facebook.com/events/...) if the extractor does not work.
